clar2wasm/src/cost.rs

```diff
@@ -0,0 +1,735 @@
+pub trait SimpleWordCharge: SimpleWord {
+    /// Generate cost tracking code for this word.
+    ///
+    /// See [`ChargeGenerator::charge`] for more details.
+    fn charge<C: ChargeGenerator>(
+        &self,
+        generator: &C,
+        instrs: &mut InstrSeqBuilder,
+        n: impl Into<Scalar>,
+    ) {
+        generator.charge(instrs, self.name(), n);
+    }
+}
+impl<Word: SimpleWord> SimpleWordCharge for Word {}
+
+/// Extension trait allowing for `ComplexWord`s to generate cost tracking code
+/// during traversal.
+// TODO: use this in the complex word traversal
+#[allow(dead_code)]
+pub trait ComplexWordCharge: ComplexWord {
+    /// Generate cost tracking code for this word.
+    ///
+    /// See [`ChargeGenerator::charge`] for more details.
+    fn charge<C: ChargeGenerator>(
+        &self,
+        generator: &C,
+        instrs: &mut InstrSeqBuilder,
+        n: impl Into<Scalar>,
+    ) {
+        generator.charge(instrs, self.name(), n);
+    }
+}
+impl<Word: ComplexWord> ComplexWordCharge for Word {}
+
+/// Generators of cost tracking code.
+pub trait ChargeGenerator {
+    /// The cost tracking context. Only present if charging code should be emitted.
+    fn cost_context(&self) -> Option<&ChargeContext>;
+
+    /// Generate code that charges the appropriate cost for the given word.
+    ///
+    /// `n` is a scaling factor that depends on the word being charged, but can only be known
+    /// during traversal. The value *must* be either a `u32` or a `LocalId` representing a local
+    /// with type `I32`.
+    /// If the word has a constant cost, the value will be ignored. This is useful in words where
+    /// the cost is known to be constant during traversal.
+    ///
+    /// Code will be generated under the following conditions:
+    ///
+    /// - The cost of the word is known
+    /// - [`CostTrackingContext::should_emit`] is `true`
+    fn charge(&self, instrs: &mut InstrSeqBuilder, word_name: ClarityName, n: impl Into<Scalar>) {
+        if let Some(ctx) = self.cost_context() {
+            if let Some(cost) = clar3::WORD_COSTS.get(&word_name) {
+                ctx.emit(instrs, cost, n);
+            }
+        }
+    }
+}
+
+impl ChargeGenerator for WasmGenerator {
+    fn cost_context(&self) -> Option<&ChargeContext> {
+        self.cost_context.as_ref()
+    }
+}
+
+/// A 32-bit unsigned integer to be resolved at either compile-time or run-time.
+#[derive(Clone, Copy)]
+pub enum Scalar {
+    Compile(u32),
+    Run(LocalId),
+}
+
+impl From<u32> for Scalar {
+    fn from(n: u32) -> Self {
+        Self::Compile(n)
+    }
+}
+
+// NOTE: We don't guarantee that the `LocalId` is of `ValType::I32`. We do this
+//       to avoid having to introduce `Module` in `ChargeContext`, and avoid
+//       having to handle an error condition originating from programmer mistake.
+//       This means the implementer should be extra careful to ensure the local
+//       is the correct type.
+impl From<LocalId> for Scalar {
```

Comment:
Not convinced by this. Someone coming to this codebase having no idea of this comment and seeing that this `impl` exist shouldn't have to go to the implementation to know about such a strong caveat.

Reply:
Since this is programmer error, and technically the compiler cannot recover from such a mistake, would a `panic!` on the type not being `I32` suffice? I ask this because, as the comment says, I would like to avoid adding an error path.

Reply:
Since it seems like I can't respond to the next comment, I'll respond to it here:

While the suggestion does indeed work - I would personally really dislike introducing (yet another) a dynamic dispatch layer. Couldn't we just do as I suggested previously in the future?

Reply:
Not if this would panic at runtime.
What is wrong in adding an error path? Forcing a dev to deal with a potential error is great for maintenance :/

Reply:
Error path added. Fortunately I could contain it to the trait impl.